test/test_nested_tensor_masking.py

```diff
@@ -39,312 +45,995 @@ def test_gen_nested_tensor(self):
-                               lambda: nt.nested_tensor_from_tensor_mask(tensor, mask))
+        tensor, mask = a.to_tensor_mask(mask_dim=2)
+        self.assertEqual(tensor, torch.tensor([[[1], [2], [3]]]))
+        self.assertEqual(mask, torch.tensor([[True, True, True]]))
+        check_tensor_options(self, tensor, a)
 
-    def test_nested_tensor_from_tensor_mask_dim_1(self):
-        original_nt = nt.nested_tensor([
-            nt.nested_tensor([
+        a = nt.nested_tensor([
                 nt.nested_tensor([
-                    torch.tensor([[1, 2], 
-                                [3, 4]])
+                    torch.tensor([1])
                 ]),
-            ]),
-            nt.nested_tensor([
                 nt.nested_tensor([
-                    torch.tensor([[5, 6],
-                                [7, 8]])
+                    torch.tensor([2])
                 ]),
+                nt.nested_tensor([
+                    torch.tensor([3])
+                ])
             ])
+        tensor, mask = a.to_tensor_mask()
+        self.assertEqual(tensor, torch.tensor([[[1]], [[2]], [[3]]]))
+        self.assertEqual(mask, torch.tensor(True))
+        check_tensor_options(self, tensor, a)
+
+        tensor, mask = a.to_tensor_mask(mask_dim=2)
+        self.assertEqual(tensor, torch.tensor([[[1]], [[2]], [[3]]]))
+        self.assertEqual(mask, torch.tensor([[True], [True], [True]]))
+        check_tensor_options(self, tensor, a)
+
+    def test_multi_tensor2(self):
+        a = nt.nested_tensor([
+                nt.nested_tensor([
+                    nt.nested_tensor([
+                        torch.tensor([[1, 2, 3, 4],
+                                      [5, 6, 7, 8]], dtype=torch.bfloat16, device='cuda', requires_grad=True)
+                    ]),
+                    nt.nested_tensor([
+                        torch.tensor([[0, 0], [3, 4]], dtype=torch.bfloat16, device='cuda', requires_grad=True)
+                    ]),
+                    nt.nested_tensor([
+                        torch.tensor([[1]], dtype=torch.bfloat16, device='cuda', requires_grad=True)
+                    ]),
+                ])
+            ])
+
+        expected_t = torch.tensor([[
+            [[[1, 2, 3, 4],
+              [5, 6, 7, 8]]],
+            [[[0, 0, 0, 0],
+              [3, 4, 0, 0]]],
+            [[[1, 0, 0, 0],
+              [0, 0, 0, 0]]],
+        ]])
+
+        expected_m = torch.tensor([[
+            [[[ True,  True,  True,  True],
+              [ True,  True,  True,  True]]],
+            [[[ True,  True, False, False],
+              [ True,  True, False, False]]],
+            [[[ True, False, False, False],
+              [False, False, False, False]]]]])
+
+        tensor, mask = a.to_tensor_mask()
+        self.assertEqual(expected_t, tensor)
+        self.assertEqual(expected_m, mask)
+        check_tensor_options(self, tensor, a)
+
+    def test_mask_dim_too_small_error(self):
+        a = nt.nested_tensor([
```

Comment:
Do you also want to then add the expected correct result or group together the tests for this specific example?

Reply:
like global test cases that are being used in different tests? i thought about it but didn't think it has enough value. I have the same input cases in very few test cases. I'm trying to have a unique case scenario per unit test - that way we can immediately see what is covered and what is broken.